libp2p/transports/mocktransport.nim

```diff
@@ -0,0 +1,59 @@
+{.push raises: [Defect].}
+
+import chronos
+import transport,
+       ../multiaddress,
+       ../stream/connection
+
+type
+  StartMock* = proc(addrs: seq[MultiAddress]): Future[void] {.gcsafe.}
+
+  StopMock* = proc(): Future[void] {.gcsafe.}
+
+  AcceptMock* = proc(): Future[Connection] {.gcsafe.}
+
+  MockTransport* = ref object of Transport
+    startMoc*: StartMock
+    stopMoc*: StopMock
+    acceptMoc*: AcceptMock
+
+const STARTMOCK_DEFAULT =
+  proc(addrs: seq[MultiAddress]) {.async.} =
+    discard
+
+const STOPMOCK_DEFAULT =
+  proc() {.async.} =
+    discard
+
+const ACCEPTMOCK_DEFAULT =
```

Comment:
I find this technique a bit weird, but maybe it's just me

Why not just create a mock directly in the test, that inherits from Transport and actually use inheritance instead of theses callback?
(I should note that I'm not a big fan of OOP generally, but since we're already inheriting, we might as well use it)

From what I can tell, we only need an optional exception to raise in start, so you can create a MockTransport that just takes a exception and raise it if it's not nil
Unless you planned to use the mock more thoroughly later?

Reply:
> I find this technique a bit weird, but maybe it's just me

At the end of the day, it's a test, and does not have an impact on the library.

> Why not just create a mock directly in the test, that inherits from Transport

This is quite cumbersome and would need to be done for each test that needs a mock.

> Unless you planned to use the mock more thoroughly later?

That was the more the point, yes. It creates a mock that allows the `start/stop/accept` procs to be injected. We can use this mock to test anything that uses transports with a unit test oriented approach, eg the switch.

Reply:
> At the end of the day, it's a test, and does not have an impact on the library.

Ok, but it's in the libp2p folder, not tests
Right now the limit between the public Api and internal stuff is a bit blurry, but I though you meant this to be a public thing

For instance we have a public mocker for name resolution, with the idea that end applications can also use it for their unit test

Could you just move it to tests folder then, if it's "private"? As long as the filename doesn't start with a "t", it shouldn't be tested

Indeed, if it's just for tests, you can used whatever you want :)

Reply:
> Could you just move it to tests folder then, if it's "private"? As long as the filename doesn't start with a "t", it shouldn't be tested

Moved :)